src/wallet/wallet.cpp

```diff
@@ -2586,209 +2584,159 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
-                        change_position->nValue += extraFeePaid;
-                        nFeeRet -= extraFeePaid;
-                    }
-                    break; // Done, enough fee included.
-                }
-                else if (!pick_new_inputs) {
-                    // This shouldn't happen, we should have had enough excess
-                    // fee to pay for the new output and still meet nFeeNeeded
-                    // Or we should have just subtracted fee from recipients and
-                    // nFeeNeeded should not have changed
-                    strFailReason = _("Transaction fee and change calculation failed").translated;
-                    return false;
-                }
+            nBytes = CalculateMaximumSignedTxSize(CTransaction(txNew), this, coin_control.fAllowWatchOnly);
+            if (nBytes < 0) {
+                strFailReason = _("Signing transaction failed").translated;
+                return false;
+            }
 
+            CAmount fee_needed = GetMinimumFee(*this, nBytes, coin_control, &feeCalc);
+            if (feeCalc.reason == FeeReason::FALLBACK && !m_allow_fallback_fee) {
+                // eventually allow a fallback fee
+                strFailReason = _("Fee estimation failed. Fallbackfee is disabled. Wait a few blocks or enable -fallbackfee.").translated;
+                return false;
+            }
+            if (nFeeRet < fee_needed) {
+                nFeeRet = fee_needed;
                 // Try to reduce change to include necessary fee
                 if (nChangePosInOut != -1 && nSubtractFeeFromAmount == 0) {
-                    CAmount additionalFeeNeeded = nFeeNeeded - nFeeRet;
                     std::vector<CTxOut>::iterator change_position = txNew.vout.begin()+nChangePosInOut;
-                    // Only reduce change if remaining amount is still a large enough output.
-                    if (change_position->nValue >= MIN_FINAL_CHANGE + additionalFeeNeeded) {
-                        change_position->nValue -= additionalFeeNeeded;
-                        nFeeRet += additionalFeeNeeded;
-                        break; // Done, able to increase fee from change
+                    change_position->nValue -= nFeeRet;
+                    // If the change is now dust, get rid of it
+                    if (IsDust(*change_position, discard_rate))
+                    {
+                        nChangePosInOut = -1;
+                        nFeeRet += change_position->nValue;
+                        txNew.vout.erase(change_position);
                     }
                 }
+                // Reduce output values for subtractFeeFromAmount
+                if (nSubtractFeeFromAmount != 0) {
+                    int i = 0;
+                    bool fFirst = true;
+                    for (const auto& recipient : vecSend)
+                    {
+                        if (i == nChangePosInOut) {
+                            ++i;
```

Comment:
Is this missing a `continue` here?

Reply:
No. What this is doing is that we are skipping the index that is reserved for the change output. Doing `continue` would skip the current recipient which we definitely don't want to do.

Reply:
Got it. I was irritated because of seeing in this case `++i` was called twice in the loop.